### PR TITLE
Use PLAIN as default SASL if not specified

### LINKED ISSUE #2117: Old golang kafka source did not require type secretKeyRef for SASL but seems to require now?
**Describe the bug**
Old golang kafka source did not require type secretKeyRef for SASL but seems to require now?

 **Expected behavior**
Existing kafka source should work without changes

 **To Reproduce**
Example Kafka Source that worked before:

```
apiVersion: sources.knative.dev/v1beta1
kind: KafkaSource
metadata:
  name: kafka-es10
spec:
  consumerGroup: kafkaes10j27c
  bootstrapServers:
  - kafka-1.mh-lbnyvywmvwwvpcmssqgl-4c201a12d7add7c99d2b22e361c6f175-0000.us-south.containers.appdomain.cloud:9093
  topics:
  - e2et10
  net:
    sasl:
      enable: true
      user:
        secretKeyRef:
          name: kafka-secret-es
          key: user
      password:
        secretKeyRef:
          name: kafka-secret-es
          key: password
    tls:
      enable: true
  sink:
    ref:
      apiVersion: v1
      kind: Service
      name: hello-display
```

With this YAML setup I got no events - I saw in kafka data-plane logs:

```
{"@timestamp":"2022-04-21T18:44:04.623Z","@version":"1","message":"Failed to reconcile [onNewEgress] egress  id=077c039a-34fb-4a51-a214-94298dd06f5b consumerGroup=kafkaes10j27c destination=http://hello-display.k9.svc.cluster.local/","logger_name":"dev.knative.eventing.kafka.broker.core.reconciler.impl.ResourcesReconcilerImpl","thread_name":"vert.x-eventloop-thread-0","level":"ERROR","level_value":40000,"stack_trace":"io.vertx.core.impl.NoStackTraceThrowable: Security protocol SASL_SSL: invalid SASL mechanism, expected SCRAM-SHA-256, SCRAM-SHA-512 or PLAIN got null\n","id":"077c039a-34fb-4a51-a214-94298dd06f5b","consumerGroup":"kafkaes10j27c","destination":"http://hello-display.k9.svc.cluster.local/"}
```

Kafak source status was showing OK:

```
k get kafkasources.sources.knative.dev ; date
NAME             TOPICS           BOOTSTRAPSERVERS                                                                                                        READY   REASON   AGE
kafka-es10       ["e2et10"]       ["kafka-1.mh-lbnyvywmvwwvpcmssqgl-4c201a12d7add7c99d2b22e361c6f175-0000.us-south.containers.appdomain.cloud:9093"]      True             3h15m
```

After adding type to Kafka source I have got all events in hello-display sink:

```
      type:
        secretKeyRef:
          name: kafka-secret-es
          key: type
```

 **Knative release version**

1.4 with V2

 **Additional context**
Add any other context about the problem here such as proposed priority

### PR DESCRIPTION
**Pull request by aslom**
Fixes #2117

## Proposed Changes

- Use PLAIN as default SASL if not specified

### CODE REVIEW

File: control-plane/pkg/security/secret.go [hunk 1]
```diff
@@ -96,12 +96,12 @@ func secretData(data map[string][]byte) kafka.ConfigOption {
 func saslConfig(protocol string, data map[string][]byte) kafka.ConfigOption {
 	return func(config *sarama.Config) error {
 
-		// Supported mechanism SASL/PLAIN or SASL/SCRAM.
+		// Supported mechanism SASL/PLAIN (default if not specified) or SASL/SCRAM.
+		saslMechanism := SaslPlain
 		givenSASLMechanism, ok := data[SaslMechanismKey]
-		if !ok {
-			return fmt.Errorf("[protocol %s] SASL mechanism required (key: %s)", protocol, SaslMechanismKey)
```
Comment (matzew) [inline control-plane/pkg/security/secret.go:102]: I like this error to go away


can we add a bit of tests to `secret_test.go` ?

File: data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/security/KubernetesCredentials.java [hunk 2]
```diff
@@ -155,14 +155,15 @@ public String SASLMechanism() {
     if (SASLMechanism == null) {
       final var SASLMechanism = secret.getData().get(SASL_MECHANISM);
       if (SASLMechanism == null) {
-        return null;
+        this.SASLMechanism  = "PLAIN"; //default when not specified
```
Comment (matzew) [inline data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/security/KubernetesCredentials.java:158]: could we add some in `KubernetesCredentialsTest.java`?

File: control-plane/pkg/security/secret.go (hunk 1 above)
Comment (aslom) [reply control-plane/pkg/security/secret.go:102]: Yes going to add tests

File: data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/security/KubernetesCredentials.java (hunk 2 above)
Comment (aslom) [reply data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/security/KubernetesCredentials.java:158]: Yes going to add tests

**Comment by aslom** [comment]
I am not able to run Java data plane unit tests - I tried with main branch (without any of my modifications)and I am getting similiar errors https://github.com/knative-sandbox/eventing-kafka-broker/issues/2143

**Review by pierDipi — COMMENTED** [review]
is it possible to add a reconciler test in https://github.com/knative-sandbox/eventing-kafka-broker/blob/main/control-plane/pkg/reconciler/source/v2/sourcev2_test.go?

**Comment by aslom** [comment]
@pierDipi do you have example of existing PLAIN type SASL setup for Kafka source? Would it be with option that add SAL with and without PLAIN typ for  NewSource from https://github.com/knative-sandbox/eventing-kafka-broker/blob/971a65528c3f42ea883d08ef07ce125aa3ec28ac/control-plane/pkg/reconciler/testing/objects_source.go#L45

**Comment by aavarghese** [comment]
@aslom maybe something like this is helpful?

https://github.com/knative-sandbox/eventing-kafka-broker/blob/main/control-plane/pkg/reconciler/consumergroup/consumergroup_test.go#L511

**Comment by aslom** [comment]
@aavarghese @pierDipi how do you run reconciler tests from command line just for SourceV2? go test?

**Comment by matzew** [comment]
> do you have example of existing PLAIN type SASL setup for Kafka source?

@aslom below is a snippet I recently used to test `PLAIN` with the `SASL_SSL`:

```yaml
apiVersion: sources.knative.dev/v1beta1
kind: KafkaSource
metadata:
  name: my-fancy-kafka-source
spec:
  bootstrapServers:
   - matze-test.kafka.cloud.com:443
  topics:
   - my-fancy
  net:
    sasl:
      enable: true
      user:
        secretKeyRef:
          name: the-kafka-sasl-secret
          key: user
      password:
        secretKeyRef:
          name: the-kafka-sasl-secret
          key: password
      type:
        secretKeyRef:
          name: the-kafka-sasl-secret
          key: saslType
    tls:
      enable: true
  sink:
    ref:
      apiVersion: serving.knative.dev/v1
      kind: Service
      name: k8s-display
```

And the `Secret` goes here:

```bash
kubectl create secret --namespace default generic the-kafka-sasl-secret \
  --from-literal=protocol=SASL_SSL \
  --from-literal=saslType="PLAIN" \
  --from-literal=sasl.mechanism=PLAIN \
  --from-literal=user="franz" \
  --from-literal=password="franz"
```

**Comment by aslom** [comment]
@matzew Great. That looks like the same Kafka source config I use for testing and that is what I implemented in recnciler tests with or without type PLAIN (BTW: it seems "protocol" and "sasl.mechanism" key are not needed in secret?)

The Kafka sources and secrets I use for testing with IBM Event Streams: 

```
apiVersion: sources.knative.dev/v1beta1
kind: KafkaSource
metadata:
  name: kafka-es10
spec:
  consumerGroup: kafkaes10j27c
  bootstrapServers:
  - kafka-1.mh-lbnyvywmvwwvpcmssqgl-4c201a12d7add7c99d2b22e361c6f175-0000.us-south.containers.appdomain.cloud:9093
  topics:
  - e2et10
  net:
    sasl:
      enable: true
      user:
        secretKeyRef:
          name: kafka-secret-es
          key: user
      password:
        secretKeyRef:
          name: kafka-secret-es
          key: password
    tls:
      enable: true
  sink:
    ref:
      apiVersion: v1
      kind: Service
      name: hello-display
```

and with type

```
apiVersion: sources.knative.dev/v1beta1
kind: KafkaSource
metadata:
  name: kafka-es10
  annotations:
    autoscaling.knative.dev/class: keda.autoscaling.knative.dev
    autoscaling.knative.dev/minScale: "1"
    autoscaling.knative.dev/maxScale: "10"
    keda.autoscaling.knative.dev/pollingInterval: "30"
    keda.autoscaling.knative.dev/cooldownPeriod: "30"
    keda.autoscaling.knative.dev/kafkaLagThreshold: "10"
spec:
  consumerGroup: kafkaes10j27c
  bootstrapServers:
  - kafka-1.mh-lbnyvywmvwwvpcmssqgl-4c201a12d7add7c99d2b22e361c6f175-0000.us-south.containers.appdomain.cloud:9093
  topics:
  - e2et10
  net:
    sasl:
      enable: true
      type:
        secretKeyRef:
          name: kafka-secret-es
          key: type
      user:
        secretKeyRef:
          name: kafka-secret-es
          key: user
      password:
        secretKeyRef:
          name: kafka-secret-es
          key: password
    tls:
      enable: true
  sink:
    ref:
      apiVersion: v1
      kind: Service
      name: hello-display
```

with secret created:

```
kubectl create secret generic kafka-secret-es10a --from-literal=type=PLAIN --from-literal=user=token --from-literal=password=VTJHYPqy9...
```

File: data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/security/KubernetesCredentials.java (hunk 2 above)
Comment (aslom) [reply data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/security/KubernetesCredentials.java]: Added tests

**Review by pierDipi — APPROVED** [review]
/lgtm
/approve

**Pull request merged**